pynav/filters.py

```diff
@@ -365,6 +411,342 @@ def _get_cost_and_info(
+            if self.sigma_mean:
+                x_mean = x.state.copy()
+                x_mean.value = mean_state(x_propagated, w)
+            else:
+                x_mean = self.process_model.evaluate(x.state, u, dt)
+
+            # Compute covariance
+            P_new = np.zeros(x.covariance.shape)
+            for i in range(n_sig):
+                err = x_mean.minus(x_propagated[i])
+                err = err.reshape((-1, 1))
+                P_new += w[i] * err @ err.T
+
+            x.state = x_mean
+            x.covariance = P_new
+            x.symmetrize()
+            x.state.stamp += dt
+
+        return x
+
+    def correct(
+        self,
+        x: StateWithCovariance,
+        y: Measurement,
+        u: Input,
+        reject_outlier: bool = None,
+    ) -> StateWithCovariance:
+        """
+        Fuses an arbitrary measurement to produce a corrected state estimate.
+
+        Parameters
+        ----------
+        x : StateWithCovariance
+            The current state estimate.
+        u: StampedValue
+            Most recent input, to be used to predict the state forward
+            if the measurement stamp is larger than the state stamp.
+        y : Measurement
+            Measurement to be fused into the current state estimate.
+        reject_outlier : bool, optional
+            Whether to apply the NIS test to this measurement, by default None,
+            in which case the value of `self.reject_outliers` will be used.
+
+        Returns
+        -------
+        StateWithCovariance
+            The corrected state estimate
+        """
+        # Make copy to avoid modifying the input
+        x = x.copy()
+
+        if x.state.stamp is None:
+            x.state.stamp = y.stamp
+
+        # Load default outlier rejection option
+        if reject_outlier is None:
+            reject_outlier = self.reject_outliers
+
+        # If measurement stamp is later than state stamp, do prediction step
+        # until current time.
+        if y.stamp is not None:
+            dt = y.stamp - x.state.stamp
+            if dt < 0:
+                raise RuntimeError(
+                    "Measurement stamp is earlier than state stamp"
+                )
+            elif u is not None:
+                x = self.predict(x, u, dt)
+
+        P = x.covariance
+        R = y.model.covariance(x.state)
+
+        n_x = x.state.dof
+        n_y = y.value.size
+
+        if (n_x, self.method) in self._sigmapoint_cache:
+            unit_sigmapoints, w = self._sigmapoint_cache[(n_x, self.method)]
+        else:
+            unit_sigmapoints, w = generate_sigmapoints(n_x, self.method)
+            self._sigmapoint_cache[(n_x, self.method)] = (unit_sigmapoints, w)
+
+        P_sqrt = np.linalg.cholesky(P)
+        sigmapoints = P_sqrt @ unit_sigmapoints
+
+        n_sig = w.size
+
+        y_check = y.model.evaluate(x.state)
+
+        if y_check is not None:
+
+            y_propagated = np.zeros((n_sig, n_y))
```

Comment:
```suggestion
            sigmapoints = sigmapoints.T
            y_propagated = [y.model.evaluate(x.state.plus(sp[0:n_x])) for sp in sigmapoints]
```
perhaps it would be best if `generate_sigmapoints` produced a [num_sigmaponts x N] array instead of a `[N x num_sigmapoints]`

anyways. by using list comprehensions when appropriate, you can get some speedup

Reply:
honestly this comment is mainly FYI, not sure if worth implementing.